### PR TITLE
Rewrite sparse dmatrix using callbacks.

### PR DESCRIPTION
**Pull request by trivialfis**
- Implemented caching with iterative DMatrix style callback, with simple, deterministic, and lock-free async fetching.  Writing to the cache, however, is sequential.
- The cache file is named by pointer address, should be able to avoid most of the collisions.
- The concatenation of ellpack page happens in the gradient sampler instead of data loading.
- `(gpu_)page_size` is removed.  Now the size of each binary block is entirely determined by the batch size provided by the user.

Part of https://github.com/dmlc/xgboost/pull/7070 .  This PR handles the internal implementation of external memory, the function is not exposed to Python yet.  High level tests are written with custom iterators without dmlc-core parser so they are still at the original PR.

### CODE REVIEW

**Comment by trivialfis** [comment]
For example usage, there's a C demo in the original PR.

**Comment by trivialfis** [comment]
Note:

The rewrite is about following things:

- Reduce dependency on dmlc parsers and provide an interface for user to load data by themselves.
- Remove dependency on threaded iterator and IO queue.
- Make sure the number of pages in memory is bounded.
- Make sure the cache can not be violated.
- Provide an interface for internal algorithm the process data asynchronously.

File: src/data/sparse_page_source.h [hunk 1]
```diff
@@ -1,564 +1,359 @@
...
-        new SparseBatchIteratorImpl<ExternalMemoryPrefetcher<CSCPage>, CSCPage>(
-            external_prefetcher_.get()));
-    return BatchSet<CSCPage>(begin_iter);
+  SparsePageSource& operator++() final {
+    TryLockGuard guard{single_threaded_};
+    count_++;
+    if (cache_info_->written) {
+      at_end_ = (count_ == n_batches_);
+    } else {
+      at_end_ = !iter_.Next();;
```
[the remark this answers is not in the document]
Comment (hcho3) [reply src/data/sparse_page_source.h:250]: 💦

**Comment by trivialfis** [comment]
@RAMitchell @hcho3 we still have some more tests we need to do in order to get it useful.  For example we need to test whether the page might go out of scope in predictor before the possibly async prediction is finished.  I will defer those into future PRs.

**Review by RAMitchell — APPROVED** [review]

**Comment by RAMitchell** [comment]
Looks good, we will now have a much more flexible external memory implementation, supporting date iterators in other languages, and more easily extending internal data structures to work with external memory.

Good to see lots of tests also.

**Review by hcho3 — APPROVED** [review]

**Pull request merged**